File: python/packages/var/src/var/internal.py
"""Author-facing API for declaring step state — port of var/src/internal.ts (defineState)."""
from __future__ import annotations

import sys
from re import Pattern
from typing import Any, Callable, Optional

from var_core.registry import Registry, add_step, create_registry, define_parameter_type
from var_core.step_role import StepKind

# ---------------------------------------------------------------------------
# Module-level mutable builder state (mirrors the module-scope vars in internal.ts)
# ---------------------------------------------------------------------------

_steps: list[dict[str, Any]] = []
# One factory per step-file; keyed by the FACTORY's __code__.co_filename
_context_factories_by_file: dict[str, Callable[[], Any]] = {}
_custom_types: list[dict[str, Any]] = []


# ---------------------------------------------------------------------------
# Public API
# ---------------------------------------------------------------------------


def define_state(
    factory: Optional[Callable[[], Any]] = None,
    param_types: Optional[dict[str, dict[str, Any]]] = None,
) -> tuple[
    Callable[[str], Callable[[Callable], Callable]],
    Callable[[str], Callable[[Callable], Callable]],
]:
    """Register *factory* as the context-state constructor for its step file.

    *factory* is optional: a step file whose steps are pure (nothing to
    arrange, nothing to evolve) calls ``define_state()`` bare and its handlers
    receive an empty ``dict`` as state.

    Returns ``(stimulus, sensor)`` — each is a decorator factory:
    ``@stimulus("expression")`` registers the decorated function as a step.

    Source location is captured from the decorated function's ``__code__``
    attributes (``co_filename`` / ``co_firstlineno``).

    Raises ``RuntimeError`` if called more than once for the same source file
    (keyed by *factory*'s ``co_filename``, or the caller's file when *factory*
    is omitted), mirroring ``internal.ts``.
    """
    if factory is None:
        # No factory code object to key by — use the calling step file itself,
        # which is where an inline factory would have been defined anyway.
        source_file: str = sys._getframe(1).f_code.co_filename
        factory = dict
    else:
        source_file = factory.__code__.co_filename
    if source_file in _context_factories_by_file:
        raise RuntimeError(
            f"defineState() called more than once in {source_file}"
        )
    _context_factories_by_file[source_file] = factory

    if param_types:
        for name, defn in param_types.items():
            _custom_types.append(
                {
                    "name": name,
                    "regexp": defn["regexp"],
                    "parse": defn.get("parse"),
                    "format": defn.get("format"),
                }
            )

    def _make_decorator(kind: StepKind) -> Callable[[str], Callable[[Callable], Callable]]:
        def decorator_factory(expression: str) -> Callable[[Callable], Callable]:
            def decorator(fn: Callable) -> Callable:
                _steps.append(
                    {
                        "expression": expression,
                        "source_file": fn.__code__.co_filename,
                        "source_line": fn.__code__.co_firstlineno,
                        "handler": fn,
                        "kind": kind,
                    }
                )
                return fn

            return decorator

        return decorator_factory

    return _make_decorator("stimulus"), _make_decorator("sensor")


def context_factory() -> Callable[[str], Any]:
    """Return a callable ``(step_file: str) -> state`` that invokes the
    registered factory for *step_file*, or ``{}`` if none is registered.

    Mirrors ``contextFactory()`` in ``internal.ts``.
    """
    # Snapshot at call time (mirrors TS semantics where the map is closed over)
    factories = dict(_context_factories_by_file)

    def _invoke(step_file: str) -> Any:
        f = factories.get(step_file)
        return f() if f is not None else {}

    return _invoke


def build_registry() -> Registry:
    """Build and return a ``Registry`` from the accumulated steps and custom types.

    Mirrors ``buildRegistry()`` in ``internal.ts``: custom parameter types are
    registered first, then steps are compiled in registration order.
    """
    r = create_registry()
    for t in _custom_types:
        kwargs: dict[str, Any] = {"name": t["name"], "regexp": t["regexp"]}
        if t.get("parse") is not None:
            kwargs["parse"] = t["parse"]
        if t.get("format") is not None:
            kwargs["format"] = t["format"]
        r = define_parameter_type(r, **kwargs)
    for e in _steps:
        r = add_step(
            r,
            expression=e["expression"],
            expression_source_file=e["source_file"],
            expression_source_line=e["source_line"],
            handler=e["handler"],
            kind=e["kind"],
        )
    return r


def _reset_builder() -> None:
    """Clear all accumulated state.  Use in tests between isolated scenarios."""
    global _steps, _custom_types
    _steps = []
    _context_factories_by_file.clear()
    _custom_types = []


def _custom_parameter_types() -> list[dict[str, str]]:
    """Conformance-harness accessor: the custom parameter types accumulated by
    ``define_state`` since the last ``_reset_builder``, projected to the
    ``{"name", "regexp"}`` wire shape ``to_registry_artifact`` serializes.

    ``regexp`` is the bare pattern source (``re.Pattern.pattern`` or the string
    as authored — no flags/delimiters), the cross-port convention every
    language's registry golden uses. Internal-only, mirrors the TS
    ``_customParameterTypes``.
    """
    out: list[dict[str, str]] = []
    for t in _custom_types:
        rx = t["regexp"]
        if isinstance(rx, Pattern):
            rx = rx.pattern
        elif not isinstance(rx, str):
            raise TypeError(
                f"parameter type {t['name']!r}: regexp lists are not supported by the "
                "conformance projection yet"
            )
        out.append({"name": t["name"], "regexp": rx})
    return out

File: ruby/packages/var/lib/oselvar/var/internal.rb
# frozen_string_literal: true

require 'oselvar/var/core'

module Oselvar
  module Var
    # The module-scope step-registration accumulator behind steps() → [param,
    # stimulus, sensor]. Mirrors @oselvar/var's internal.ts. A step file, when
    # loaded, calls steps() once and registers into the accumulators here; the
    # runner/harness then reads them via build_registry / context_factory.
    module Internal
      @steps = []
      @context_factories_by_file = {}
      @custom_types = []

      class << self
        # Register a file's state factory and return [param, stimulus, sensor].
        # +factory+ is a callable (or nil for empty state); +source_file+ keys
        # the per-file context factory. Raises if called twice for one file.
        def register(factory, source_file)
          raise "steps() called more than once in #{source_file}" if @context_factories_by_file.key?(source_file)

          @context_factories_by_file[source_file] = factory || -> { {} }
          [param_registrar, registrar('stimulus'), registrar('sensor')]
        end

        # (step_file) -> state: invoke the file's factory, or {} if none.
        def context_factory
          factories = @context_factories_by_file.dup
          lambda do |step_file|
            factory = factories[step_file]
            factory ? factory.call : {}
          end
        end

        # Build a Core::Registry: custom parameter types first (so expressions
        # can reference them), then steps in registration order.
        def build_registry
          registry = Core::Registries.create_registry
          @custom_types.each do |type|
            registry = Core::Registries.define_parameter_type(
              registry, name: type[:name], regexp: type[:regexp], parse: type[:parse], format: type[:format]
            )
          end
          @steps.each do |step|
            registry = Core::Registries.add_step(
              registry,
              expression: step[:expression],
              expression_source_file: step[:source_file],
              expression_source_line: step[:source_line],
              handler: step[:handler],
              kind: step[:kind]
            )
          end
          registry
        end

        # Clear all accumulated state (between isolated runs / harness bundles).
        def reset_builder
          @steps = []
          @context_factories_by_file = {}
          @custom_types = []
        end

        # Conformance-harness accessor: custom parameter types projected to the
        # {"name","regexp"} wire shape. `regexp` is the bare source (Regexp#source
        # or the string as authored) — the cross-port convention.
        def custom_parameter_types
          @custom_types.map do |type|
            regexp = type[:regexp]
            regexp = regexp.source if regexp.is_a?(Regexp)
            unless regexp.is_a?(String)
              raise "parameter type #{type[:name].inspect}: regexp arrays are not supported " \
                    'by the conformance projection yet'
            end
            { 'name' => type[:name], 'regexp' => regexp }
          end
        end

        private

        def param_registrar
          lambda do |name, regexp, parse: nil, format: nil|
            @custom_types << { name: name, regexp: regexp, parse: parse, format: format }
            nil
          end
        end

        def registrar(kind)
          lambda do |expression, &handler|
            file, line = handler.source_location
            @steps << {
              expression: expression, source_file: file, source_line: line,
              handler: handler, kind: kind
            }
            nil
          end
        end
      end
    end
  end
end
